src/aspire/classification/rir_class2d.py

```diff
@@ -0,0 +1,577 @@
+                f"Requested {self.n_nbor} self.n_nbor, but only {n_im} images. Setting self.n_nbor={n_im-1}."
+            )
+            n_nbor = n_im - 1
+
+        concat_coeff = np.concatenate((coeff_b, coeff_b_r), axis=1)
+
+        num_batches = (n_im + batch_size - 1) // batch_size
+
+        classes = np.zeros((n_im, n_nbor), dtype=int)
+        for i in range(num_batches):
+            start = i * batch_size
+            finish = min((i + 1) * batch_size, n_im)
+            corr = np.real(
+                np.dot(np.conjugate(coeff_b[:, start:finish]).T, concat_coeff)
+            )
+            # Note legacy did not include the original image?
+            # classes[start:finish] = np.argsort(-corr, axis=1)[:, 1 : n_nbor + 1]
+            # This now does include the original image
+            # (Matches sklean implementation.)
+            # Check with Joakim about preference.
+            # I (GBW) think class[i] should have class[i][0] be the original image index.
+            classes[start:finish] = np.argsort(-corr, axis=1)[:, :n_nbor]
+
+        return classes
+
+    def _legacy_pca(self, M):
+        """
+        This is more or less the historic implementation ported
+        to Python from code calling MATLAB's `pca_y`.
+        """
+
+        # ### The following was from legacy code. Be careful wrt order.
+        M = M.T
+        u, s, v = pca_y(M, self.bispectrum_componenents)
+
+        # Contruct coefficients
+        coef_b = np.einsum("i, ij -> ij", s, np.conjugate(v))
+        coef_b_r = np.conjugate(u.T).dot(np.conjugate(M))
+
+        # Normalize
+        coef_b /= np.linalg.norm(coef_b, axis=0)
+        coef_b_r /= np.linalg.norm(coef_b_r, axis=0)
+
+        # Transpose (this code was originally F order)
+        coef_b = coef_b.T
+        coef_b_r = coef_b_r.T
+
+        return coef_b, coef_b_r
+
+    def _sk_pca(self, M):
+        # Avoiding SK directly for now,
+        #   while it is really useful, it
+        #   expects real data.
+        #   We use an extension of SK that is hacked to admit complex.
+        pca = ComplexPCA(
+            self.bispectrum_componenents,
+            copy=False,  # careful, overwrites data matrix... we'll handle the copies.
+            svd_solver="auto",  # use randomized (Halko) for larger problems
+            random_state=self.seed,
+        )
+        coef_b = pca.fit_transform(M.copy())
+        coef_b_r = pca.fit_transform(np.conjugate(M))
+
+        # M is no longer needed, hint for it to get cleaned up.
+        del M
+
+        # I'm also not sure why this norm is needed...
+        #  but it does work better with it.
+        coef_b /= np.linalg.norm(coef_b, axis=1)[:, np.newaxis]
+        coef_b_r /= np.linalg.norm(coef_b_r, axis=1)[:, np.newaxis]
+
+        return coef_b, coef_b_r
+
+    def _devel_bispectrum(self, coef):
+        coef = self.pca_basis.to_complex(coef)
+        eigvals = self.pca_basis.to_complex(self.pca_basis.eigvals).reshape(
```

Comment:
I agree, here and also where I compute the eigval magntitudes is not exactly right (there is a comment to that effect in the other area of concern).  While this isn't quite right it seems to work okay, I'll have to think about it more unless you can point me to some documentation.

Reply:
From what I recall, the eigenvalues for the complex coefficients are identical to the eigenvalues for real coefficients. In other words, the eigenvalues for the positive and negative (real) coefficients are the same for a given `(ell, k)` and those are the eigenvalues for the complex coefficients (modulo something funny going on for `ell = 0` maybe).

Reply:
Okay, just checked and I agree. I definitely overlooked that.  Because these complex values a+ai are `abs` a little later I don't think this will change results, but I will update the code to just loop and take ells>=0.

Reply:
Changed to reals, added note.

Reply:
Got it. Thanks.